DOC: Add explanation of a sparse mesh grid

timhoffm opened the pull request:
See https://github.com/numpy/numpy/pull/19561#issuecomment-886501122.

I've also changed `np.arange(-5, 5, 0.1)` to `np.linspace(-5, 5, 101)` in the example because `arange` for non-integer steps is generally not recommended.

https://numpy.org/doc/stable/reference/generated/numpy.arange.html:

> When using a non-integer step, such as 0.1, the results will often not be consistent. It is better to use numpy.linspace for these cases.

timhoffm committed:
Add explanation of a spare mesh grid.

Follow-up to #19561.

rossbar commented on numpy/lib/function_base.py:
```diff
@@ -4277,7 +4277,20 @@ def meshgrid(*xi, copy=True, sparse=False, indexing='xy'):
+        in an expression, broadcasting still leads to a fully-dimensonal
+        result array.
+
+        Example: In the second example below with a (101, 101) shaped grid,
+        the total size of the coordinate arrays ``xx`` and ``yy`` is reduced
+        from 2*101*101 = 20.402 to 101+101 = 202 values. Likewise, the
+        expression ``xx**2 + yy**2`` takes 20.401 square operations with full
+        coordinates but only 202 square operations with sparse coordinates,
+        while the resulting is the same.
+
```
I think this explanatory text would fit better directly in the examples section. That way some of these points could be illustrated directly, e.g. comparing `xx.shape` w/ both `sparse=True/False`

charris commented:
@timhoffm @rossbar  Ping.

rossbar commented on numpy/lib/function_base.py:
```diff
@@ -4351,14 +4357,27 @@ def meshgrid(*xi, copy=True, sparse=False, indexing='xy'):
+    can use the parameter ``sparse=True`` to save memory and computation time.
+    Here, the sparse coordinate grid reduces the memory from 2*101*101 = 20.402
+    to 101+101 = 202 values.  Likewise, the expression ``xx**2 + yy**2`` is
+    reduced from 20.401 square operations to only 202 square operations, but
```
I don't think this is correct. There's both the squaring and addition operations to consider: doesn't the addition require 101**2 operations?

In any case, my two cents is that this could benefit from less detail as it's easy to lose the forest through the trees here. I'd advocate for chopping out the text from `Here,`: IMO the best way to provide more detail would be to add a reference to the [2011 NumPy paper](https://arxiv.org/abs/1102.1523), which has an excellent detailed example of broadcasting with "sparse" arrays.

Just my two cents though - leaving the detail is also fine, but we should double check to make sure the numbers are correct.

timhoffm replied:
I was only referring to the square operations. I anticipated that the square operation was more expensive than addition, but that doesn't seem to be the case. So the overall gain is theoretically just a factor 3 (beacuse the sparse sqares can be neglected compared to the addition). A bit more testing also shows that a real gain only appears when x, y have order of 1000 elements or larger, which I find surprisingly high.

![grafik](https://user-images.githubusercontent.com/2836374/132109337-3200ff9d-c2e6-4c8c-b258-b4f59d326882.png)

Anyway, it's not within my scope to discuss performance characteristics. I've removed everything from "Here" as suggested.

charris commented:
close/reopen

rossbar reviewed (commented):
Oops - I only meant that the text about performance should be removed, not the example. I think the clearest way to demonstrate what the `sparse` kwarg does is to show it:

```python
>>> xx, yy = np.meshgrid(x, y)
>>> xx.shape, yy.shape
((101, 101), (101, 101))
>>> xs, ys = np.meshgrid(x, y, sparse=True)
>>> xs.shape, ys.shape
((1, 101), (101, 1))
>>> zs = np.sin(xs**2 + ys**2) / (xs**2 + ys**2)
>>> np.array_equal(z, zs, equal_nan=True)
True
>>> zs.shape
(101, 101)
```

Also, it seems that `sparse` keeps getting replaced with `spare` in the pushes.

timhoffm commented:
@rossbar Thanks, I've taken your example suggestion and adapted:

- have separate sections for full and sparse arrays. - This example shows how to use a `meshgrid` to evaluate a function. Having first a complete "full" example is serves this primary purpose without already mixing in spare logic.
- simplified the function to evaluate.
- removed the plot. - While it is nice for illustration of "evaluate a function", the plotting code and the need for a more complex function would make the example longer and harder to read.

rossbar commented:
> removed the plot. - While it is nice for illustration of "evaluate a function", the plotting code and the need for a more complex function would make the example longer and harder to read.

Personally I'm -1 on this change. This is an example of a very common use-case that demonstrates how these two fundamental packages (numpy and matplotlib) are used together. It'd be a shame to lose that, and I don't really think the 4 extra lines for matplotlib make the example difficult to follow.

If you feel strongly about removing it, I'd recommend opening a separate PR for it.

timhoffm commented:
I've reverted the plot removal, Instead I've moved it to a separate block and added a colorbar.

rossbar committed:
Fix doctest formatting for example comments.

rossbar commented:
Thanks @timhoffm !
